File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cs
public class Solution {
    public int[] TopKFrequent(int[] nums, int k) {
        var freqMap = new Dictionary<int,int>();
        foreach(int num in nums){
            freqMap[num] = freqMap.GetValueOrDefault(num, 0) + 1;
        }
        var heap = new PriorityQueue<int, int>();
        foreach(var pair in freqMap){
            heap.Enqueue(pair.Key, pair.Value);
            if(heap.Count > k){
                heap.Dequeue();
            }
        }
        var result = new List<int>();
        while(heap.Count > 0){
            result.Add(heap.Dequeue());
        }
        return result.ToArray();
    }
}

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function (nums, k) {
    const freqMap = new Map();
    for (let num of nums) {
        freqMap.set(num, (freqMap.get(num) || 0) + 1);
    }
    const buckets = Array(nums.length + 1).fill().map(() => []);
    for (let [num, freq] of freqMap.entries()) {
        buckets[freq].push(num);
    }
    const res = [];
    for (let i = buckets.length - 1; i >= 0 && res.length < k; i--) {
        for (let num of buckets[i]) {
            res.push(num);
            if (res.length === k) break;
        }
    }
    return res
};